drivers/hidparser.c

```diff
@@ -460,49 +460,25 @@ void GetValue(const unsigned char *Buf, HIDData_t *pData, long *pValue)
 	"throwing away higher-order bits" exacly means, so we try to do
 	something sensible. -PS */
 
-	rawvalue = value; /* remember this for later */
+	/* determine representation without sign bit */
+	magMax = pData->LogMax >= 0 ? pData->LogMax : -(pData->LogMax + 1);
+	magMin = pData->LogMin >= 0 ? pData->LogMin : -(pData->LogMin + 1);
 
-	/* figure out how many bits are significant */
-	range = pData->LogMax - pData->LogMin + 1;
-	if (range <= 0) {
-		/* makes no sense, give up */
-		*pValue = value;
-		return;
-	}
-	b = hibit(range-1);
+	/* calculate where the sign bit will be if needed */
+	signbit = 1L << hibit(magMax > magMin ? magMax : magMin);
```

Comment:
There's a lot packed up in there to respond to but perhaps the place to start is

> platforms where a `long` might be less than 4 bytes

the C standard, e.g, [ISO C99 draft](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) precludes that possibility in section 5.2.4.2.1 Sizes of integer types, where it says that the LONG_MAX value must be no less than +2147483647.

The two common size configurations we currently see are ILP32 (ints, longs, pointers are all 32-bit) and LP64 (int is 32-bit, but long and pointer are 64-bit).  I haven't personally encountered an ILP64 system, but I gather some may exist.

We know that `1L` is by definition the same size as `long` -- I think introducing `1ULL`, that is `unsigned long long` to the mix would add to confusion and possible errors rather than reduce it.

I think that if the rest of the NUT code were converted to use explicit sizes rather than the existing short, int, long, long long, and even some "unsigned short int" and "unsigned long long int", then it would be time to do the same here.   I also think that lacking test cases for both conforming and non-conforming (to the HID spec) UPSs there's much chance of doing that and still producing correct results.

Reply:
BTW, I just tried running test cases for GetValue, with signbit and mask being uint32_t, and then with magMin and magMax either being unsigned long or uint32_t.  In none of these cases did it produce the correct result when the program was compiled in 64-bit (LP64) mode.  I haven't done the analysis to determine where things go wrong in these cases, but I'd argue that it's more important to get the correct answer...

Reply:
Thanks for the clarifications! I wonder if a version of that makes sense as a comment in the code, and a link to the issues and PRs you made on this point, so "future-we" (generally in the community) do not try to blindly "optimize" this and break stuff unwittingly.

Regarding the latter comment, I take it as that the current solution works for all architectures you could get hands on, and "similar" implementations as well as original one did not?

Also, do you have some test code to share (maybe into same comment, so future "wannabe optimizers" can test their fixes and/or put that into unit tests eventually)? It might help confirm non-regression on other platforms to be sure :)